### PR TITLE
[tune] PBT + Memnn example

### PR DESCRIPTION
**Pull request by anthonyhsyu**
## Why are these changes needed?

Adds an example based off of the bAbI-MemNN example under Keras found at [https://keras.io/examples/babi_memnn/](https://keras.io/examples/babi_memnn/)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

File: python/ray/tune/examples/memnn_example.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+        with tarfile.open(path) as tar:
+            train_stories = get_stories(
+                tar.extractfile(challenge.format("train")))
+            test_stories = get_stories(
+                tar.extractfile(challenge.format("test")))
+
+        return train_stories, test_stories
+
+    def _build_model(self, input_shape):
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]:
```suggestion
    def build_model(self, input_shape):
    	"""Helper method for creating model"""
```

File: python/ray/tune/examples/memnn_example.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    for story, query, answer in data:
+        inputs.append([word_idx[w] for w in story])
+        queries.append([word_idx[w] for w in query])
+        answers.append(word_idx[answer])
+    return (pad_sequences(inputs, maxlen=story_maxlen),
+            pad_sequences(queries, maxlen=query_maxlen), np.array(answers))
+
+
+class MemNNModel(Trainable):
+    def _read_data(self):
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: move this outside?

File: python/ray/tune/examples/memnn_example.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/examples/memnn_example.py]: take out input_shape?

File: python/ray/tune/examples/memnn_example.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        model = self._build_model(None)  # Shape not used
+        model.compile(
+            optimizer="rmsprop",
+            loss="sparse_categorical_crossentropy",
+            metrics=["accuracy"])
+        self.model = model
+
+    def _train(self):
+        batch_size = 32
+        epochs = 1
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: this could go into `config`

File: python/ray/tune/examples/memnn_example.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+        # train
+        self.model.fit(
+            [self.inputs_train, self.queries_train],
+            self.answers_train,
+            batch_size=batch_size,
+            epochs=epochs,
+            validation_data=None,
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: maybe you can just include the validation data here?

File: python/ray/tune/examples/memnn_example.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            verbose=0)
+        _, accuracy = self.model.evaluate(
+            [self.inputs_test, self.queries_test], self.answers_test)
+        return {"mean_accuracy": accuracy}
+
+    def _save(self, checkpoint_dir):
+        file_path = checkpoint_dir + "/model"
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: `os.path.join(checkpoint_dir, "model")`

File: python/ray/tune/examples/memnn_example.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        self.model.save(file_path)
+        return file_path
+
+    def _restore(self, path):
+        # See https://stackoverflow.com/a/42763323
+        del self.model
+        self.model = load_model(path)
+
+    def _stop(self):
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: remove?

File: python/ray/tune/examples/memnn_example.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        mode="max",
+        perturbation_interval=20,
+        hyperparam_mutations={"dropout": lambda: np.random.uniform(0, 1)})
+
+    run(MemNNModel,
+        name="pbt_babi_memnn",
+        scheduler=pbt,
+        stop={"training_iteration": 50},
+        num_samples=10,
+        **{
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: no need for this

File: python/ray/tune/examples/memnn_example.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        model = Model([input_sequence, question], answer)
+        return model
+
+    def _setup(self, config):
+        self.train_stories, self.test_stories = self._read_data()
+        model = self._build_model(None)  # Shape not used
+        model.compile(
+            optimizer="rmsprop",
+            loss="sparse_categorical_crossentropy",
+            metrics=["accuracy"])
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: modify learning rate?

File: python/ray/tune/examples/memnn_example.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            [match, input_encoded_c])  # (samples, story_maxlen, query_maxlen)
+        response = Permute(
+            (2, 1))(response)  # (samples, query_maxlen, story_maxlen)
+
+        # concatenate the match matrix with the question vector sequence
+        answer = concatenate([response, question_encoded])
+
+        # the original paper uses a matrix multiplication for this reduction step.
+        # we choose to use a RNN instead.
+        answer = LSTM(32)(answer)  # (samples, 32)
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: could also change this

Comment (richardliaw) [reply python/ray/tune/examples/memnn_example.py]: actually maybe you can't change your architecture that easily

File: python/ray/tune/examples/memnn_example.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        "--smoke-test", action="store_true", help="Finish quickly for testing")
+    args, _ = parser.parse_known_args()
+    ray.init()
+
+    pbt = PopulationBasedTraining(
+        time_attr="training_iteration",
+        metric="mean_accuracy",
+        mode="max",
+        perturbation_interval=20,
+        hyperparam_mutations={"dropout": lambda: np.random.uniform(0, 1)})
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: mutate like learning rate, momentum (in rmsprop), and dropout

**Review by richardliaw — COMMENTED** [review]
@anthonyhsyu can you add this to the tune_tests under ci/jenkins_tests/?

File: python/ray/tune/examples/memnn_example.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
+"""Example training a memory neural net on the bAbI dataset.
+
+References Keras and is based off of https://keras.io/examples/babi_memnn/.
+"""
+
+from __future__ import print_function
+
+from keras.models import Sequential, Model, load_model
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: oh; as of now, looks like we should move all of this to `tf.keras`; could you do that? should just be an import swap.

File: python/ray/tune/examples/memnn_example.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+
+if __name__ == "__main__":
+    import ray
+    from ray.tune import Trainable, run
+    from ray.tune.schedulers import PopulationBasedTraining
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--smoke-test", action="store_true", help="Finish quickly for testing")
```
Comment (richardliaw) [inline python/ray/tune/examples/memnn_example.py]: The smoke-test doesn't actually seem to do anything? You should limit `training_iteration` to like 20 if `smoke_test` is True.

Also, maybe set num_samples to like 4? So that this runs better on our CI.

**Review by richardliaw — APPROVED** [review]
This looks really good. Can you post a TensorBoard graph? https://ray.readthedocs.io/en/latest/tune-usage.html#tensorboard

**Comment by anthonyhsyu** [comment]
Screen Shot 2019-09-24 at 11 21 44 AM
